### PR TITLE
fix(jax): handle DPA-2 pbc/nopbc without mapping

### PR DESCRIPTION
**Pull request by njzjz**
In the C++ API, generate the mapping for the no PBC and throw the error for PBC. Considering I forgot setting `atom_modify map yes` when testing it, others may also forget.

## Summary by CodeRabbit

- **New Features**
  - Introduced a function to determine if the model supports message passing, enhancing the model's interface.
  - Added a private member variable to facilitate message passing functionality.
  - Implemented unit tests for the `DeepPot` class, validating its functionality under various conditions.

- **Bug Fixes**
  - Improved error handling for TensorFlow function retrieval, ensuring more specific exceptions are thrown.
  - Enhanced compatibility with earlier model versions by managing exceptions related to the new message passing variable.

### CODE REVIEW

File: source/api_cc/include/DeepPotJAX.h [hunk 1]
```diff
@@ -189,6 +189,8 @@ class DeepPotJAX : public DeepPotBackend {
   std::vector<int64_t> sel;
   // number of neighbors
   int nnei;
+  // do message passing
+  bool do_message_passing;
```
[the remark this answers is not in the document]
Comment (njzjz) [reply source/api_cc/include/DeepPotJAX.h:193]: ??? why it's `int`?

File: source/api_cc/src/DeepPotJAX.cc [hunk 2]
```diff
@@ -72,7 +72,8 @@ inline TFE_Op* get_func_op(TFE_Context* ctx,
   TF_Function* func = NULL;
   find_function(func, funcs, func_name);
   if (func == NULL) {
-    throw std::runtime_error("Function " + func_name + " not found");
+    // note: the message format is used by error catching of do_message_passing
+    throw deepmd::deepmd_exception("Function " + func_name + " not found");
```
[the remark this answers is not in the document]
Comment (njzjz) [reply source/api_cc/src/DeepPotJAX.cc:76]: Reasonable.

File: source/api_cc/include/DeepPotJAX.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (njzjz) [reply source/api_cc/include/DeepPotJAX.h:193]: Shall we change it to `bool`?

File: source/api_cc/src/DeepPotJAX.cc [hunk 3]
```diff
@@ -314,6 +315,17 @@ void deepmd::DeepPotJAX::init(const std::string& model,
...
+    do_message_passing = get_scalar<bool>(ctx, "do_message_passing",
+                                          func_vector, device, status);
+  } catch (deepmd::deepmd_exception& e) {
+    // compatibile with models generated by v3.0.0rc0
+    if (e.what() == "Function do_message_passing not found") {
+      do_message_passing = false;
+    } else {
+      throw e;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (njzjz) [reply source/api_cc/src/DeepPotJAX.cc]: Fixed in https://github.com/deepmodeling/deepmd-kit/pull/4363/commits/59952fafb50b6a93828c5e9b8a8f938fbbf731a8

File: source/api_cc/src/DeepPotJAX.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (njzjz) [reply source/api_cc/src/DeepPotJAX.cc]: Fixed in Fixed in https://github.com/deepmodeling/deepmd-kit/pull/4363/commits/59952fafb50b6a93828c5e9b8a8f938fbbf731a8

File: source/api_cc/include/DeepPotJAX.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (njzjz) [reply source/api_cc/include/DeepPotJAX.h:193]: > Would you like to update the declarations in `DeepPotPT.h` and `DeepSpinPT.h` to use `bool` for `do_message_passing`?

Yes. Please open an issue for it.

**Review by wanghan-iapcm — APPROVED** [review]

**Pull request merged**